src/app/rosetta/lib/block.ml

```diff
@@ -0,0 +1,579 @@
+      ; fee_payer= `Pk "Alice"
+      ; source= `Pk "Alice"
+      ; token= Unsigned.UInt64.of_int 10
+      ; fee= Unsigned.UInt64.of_int 2_000_000_000
+      ; receiver= `Pk "Bob"
+      ; fee_token= Unsigned.UInt64.of_int 1
+      ; nonce= Unsigned.UInt32.of_int 3
+      ; amount= Some (Unsigned.UInt64.of_int 30_000)
+      ; failure_status= Some (`Applied Account_creation_fees_paid.By_no_one)
+      ; hash= "TXN_5_HASH" } ]
+end
+
+module Internal_command_info = struct
+  module Kind = struct
+    type t = [`Coinbase | `Fee_transfer] [@@deriving eq]
+  end
+
+  type t =
+    { kind: Kind.t
+    ; receiver: [`Pk of string]
+    ; fee: Unsigned.UInt64.t
+    ; token: Unsigned.UInt64.t
+    ; hash: string }
+
+  let to_operations ~user_commands ~coinbase ~block_creator_pk (t : t) :
+      Operation.t list =
+    (* We choose to represent the dec-side of fee transfers from txns from the
+     * canonical user command that created them so we are able consistently
+     * produce more balance changing operations in the mempool or a block.
+     * *)
+    let plan : 'a Op.t list =
+      match t.kind with
+      | `Coinbase ->
+          [{Op.label= `Coinbase_inc; related_to= None}]
+      | `Fee_transfer ->
+          (* Detect if a fee transfer comes from a snark worker payment by
+         * failing to find the user command that caused it *)
+          if
+            List.find user_commands ~f:(fun (info : User_command_info.t) ->
+                Unsigned.UInt64.equal info.fee t.fee
+                && Unsigned.UInt64.equal info.token t.token
+                && [%eq: [`Pk of string]] info.receiver t.receiver )
+            |> Option.is_none
+          then
+            (* In this case we make a "synthetic" fee subtraction from block
+           * creator even though this is implicit in our blocks. We need this
+           * so the balance always sums to zero. *)
+            [ {Op.label= `Fee_payer_dec; related_to= None}
+            ; {Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec} ]
+          else
+            (* Otherwise we only have the receiver inc side *)
+            (* TODO: Relate fee_payer_dec with fee_receiver_inc across
+           * user_commands and internal commands *)
+            [{Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec}]
+    in
+    Op.build ~a_eq:[%eq: [`Coinbase_inc | `Fee_payer_dec | `Fee_receiver_inc]]
+      ~plan ~f:(fun ~related_operations ~operation_identifier op ->
+        (* All internal commands succeed if they're in blocks *)
+        let status = Operation_statuses.name `Success in
+        match op.label with
+        | `Coinbase_inc ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account=
+                Some (account_id block_creator_pk Amount_of.Token_id.default)
+            ; _type= Operation_types.name `Coinbase_inc
+            ; amount= Some (Amount_of.coda coinbase)
```

Comment:
The amount here should be `coinbase - fee-transfer-amount`. `fee-transfer-amount` comes from the fee transfer that is part of the coinbase transaction

Reply:
Another approach (that I would prefer) would be to consider the fee transfer separately as its own `fee_transfer_dec`. I don't think the coinbase amount should vary; that it does is purely an implementation detail of the bundled fee transfer.